Preallocate quota for FileSystemSyncAccessHandle

szewai opened the pull request:
#### 459a045d57da232e5149f017fb41038df2a58545
<pre>
Preallocate quota for FileSystemSyncAccessHandle
<a href="https://bugs.webkit.org/show_bug.cgi?id=251642">https://bugs.webkit.org/show_bug.cgi?id=251642</a>
rdar://104980443

Reviewed by Youenn Fablet.

QuotaManager is in network process, and FileSystemSyncAcessHandle performs file operation in web process, so web process
needs to message network process when it needs to use more space. However, FileSystemSyncAccessHandle::write is sync and
supposed to be performant, which means we should unblock them if possible. This patch makes QuotaManager allocate more
capacity to FileSystemSyncAccessHandle than it requests, so that FileSystemSyncAccessHandle would not need to send a
message on each write).

The current allocation policy for requested capacity x is:
1. If x &lt;= 1MB, new capacity is 1MB.
2. If 1MB &lt; x &lt; 256MB, new capacity is 2^(log2(x) + 1).
3. If x &gt;= 256MB, new capacity is (x / 128MB + 1) * 128MB.

With this change, FileSystemSyncAcessHandle tests will start to fail if we disallow quota to increase (e.g. disallow it
to check if quota error can be thrown), because TestRunner sets origin quota to be 40KB by default, and each
FileSystemSyncAcessHandle requests at least 1MB. To fix this, this patch adds a new setQuota() function on TestRunner to
allow configuring quota for each test.

* LayoutTests/storage/filesystemaccess/resources/sync-access-handle-storage-limit.js:
* LayoutTests/storage/filesystemaccess/sync-access-handle-storage-limit-worker-expected.txt:
* LayoutTests/storage/filesystemaccess/sync-access-handle-storage-limit-worker.html:
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::requestNewCapacityForSyncAccessHandle):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setQuota):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setQuota):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
(WTR::TestController::setQuota):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate init]):
(-[TestWebsiteDataStoreDelegate requestStorageSpace:frameOrigin:quota:currentSize:spaceRequired:decisionHandler:]):
(-[TestWebsiteDataStoreDelegate setQuota:]):

Canonical link: <a href="https://commits.webkit.org/259946@main">https://commits.webkit.org/259946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/589174e9f68563e5c783b7f38c154f9b4e16ad64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115584 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175686 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110302 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16916 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6654 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98662 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115254 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12846 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95829 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40462 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94721 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27472 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8662 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28824 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9189 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5396 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14814 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48370 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6872 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10741 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
```diff
@@ -295,14 +298,21 @@ void FileSystemStorageHandle::requestNewCapacityForSyncAccessHandle(WebCore::Fil
+    if (newCapacity < defaultInitialCapacity)
+        newCapacity = defaultInitialCapacity;
+    else if (newCapacity < defaultMaxCapacityForExponentialGrowth)
+        newCapacity = pow(2, (int)std::log2(newCapacity) + 1);
```
Why not `newCapcity *= 2;`?

szewai replied:
The current heuristic tries not preallocating too much and not growing too slow.
`newCapcity *= 2` might allocate too much if the pattern is a big write followed by some small writes (e.g. 100MB -> 1MB -> 1MB). It might not differ too much if the pattern is continuous small writes (e.g. 1MB -> 1MB -> 1MB). 
We could check how it performs and make adjustment if needed.

youennf commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -327,6 +327,7 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
     }
 
     [globalWebsiteDataStoreDelegateClient() setAllowRaisingQuota:YES];
+    [globalWebsiteDataStoreDelegateClient() setQuota:40 * KB];
```
It would be better to call setQuota(40 * KB) in platform agnostic code.

youennf commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -518,6 +519,11 @@ void initializeWebViewConfiguration(const char* libraryPath, WKStringRef injecte
 
+void TestController::setQuota(uint64_t quota)
+{
+    [globalWebsiteDataStoreDelegateClient() setQuota: quota];
```
```suggestion
    [globalWebsiteDataStoreDelegateClient() setQuota:quota];
```

szewai committed:
Preallocate quota for FileSystemSyncAccessHandle
https://bugs.webkit.org/show_bug.cgi?id=251642
rdar://104980443

Reviewed by Youenn Fablet.

QuotaManager is in network process, and FileSystemSyncAcessHandle performs file operation in web process, so web process
needs to message network process when it needs to use more space. However, FileSystemSyncAccessHandle::write is sync and
supposed to be performant, which means we should unblock them if possible. This patch makes QuotaManager allocate more
capacity to FileSystemSyncAccessHandle than it requests, so that FileSystemSyncAccessHandle would not need to send a
message on each write).

The current allocation policy for requested capacity x is:
1. If x <= 1MB, new capacity is 1MB.
2. If 1MB < x < 256MB, new capacity is 2^(log2(x) + 1).
3. If x >= 256MB, new capacity is (x / 128MB + 1) * 128MB.

With this change, FileSystemSyncAcessHandle tests will start to fail if we disallow quota to increase (e.g. disallow it
to check if quota error can be thrown), because TestRunner sets origin quota to be 40KB by default, and each
FileSystemSyncAcessHandle requests at least 1MB. To fix this, this patch adds a new setQuota() function on TestRunner to
allow configuring quota for each test.

* LayoutTests/storage/filesystemaccess/resources/sync-access-handle-storage-limit.js:
* LayoutTests/storage/filesystemaccess/sync-access-handle-storage-limit-worker-expected.txt:
* LayoutTests/storage/filesystemaccess/sync-access-handle-storage-limit-worker.html:
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::requestNewCapacityForSyncAccessHandle):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setQuota):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::setQuota):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
(WTR::TestController::setQuota):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate init]):
(-[TestWebsiteDataStoreDelegate requestStorageSpace:frameOrigin:quota:currentSize:spaceRequired:decisionHandler:]):
(-[TestWebsiteDataStoreDelegate setQuota:]):

Canonical link: https://commits.webkit.org/259946@main
